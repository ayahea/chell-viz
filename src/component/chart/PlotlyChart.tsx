import * as Immutable from 'immutable';
import * as React from 'react';

// We need to use dist/plotly to avoid forcing users to do some webpack gymnastics:
// https://github.com/plotly/plotly-webpack#the-easy-way-recommended
import * as plotly from 'plotly.js/dist/plotly';

import { withDefaultProps } from '../../helper/ReactHelper';

export enum PLOTLY_CHART_TYPE {
  /** [Plotly Bar Chart](https://plot.ly/javascript/bar-charts/) */
  'bar' = 'bar',
  /** [Plotly Point Cloud](https://plot.ly/javascript/pointcloud/) */
  'pointcloud' = 'pointcloud',
  /** [Plotly Line/Scatter Chart](https://plot.ly/javascript/line-and-scatter/) */
  'scatter' = 'scatter',
  /** [Plotly Line/Scatter Chart in WebGL](https://plot.ly/javascript/line-and-scatter/) */
  'scattergl' = 'scattergl',
  /** [Plotly 3D Scatter Plot](https://plot.ly/javascript/3d-scatter-plots/) */
  'scatter3d' = 'scatter3d',
}

export interface IPlotlyData extends plotly.ScatterData {
  type: PLOTLY_CHART_TYPE | 'bar' | 'pointcloud' | 'scatter' | 'scattergl' | 'scatter3d';
}

export const defaultPlotlyConfig: Partial<Plotly.Config> = {
  displayModeBar: false,
  scrollZoom: true,
  // modeBarButtons: [['zoomOut2d', 'zoomIn2d'], ['resetScale2d', 'autoScale2d'], ['select2d', 'pan2d']],
};

export const defaultPlotlyLayout: Partial<Plotly.Layout> = {
  autosize: true,
  height: 400,
  hovermode: 'closest',
  legend: {},
  margin: {
    b: 10,
    l: 40,
    r: 10,
    t: 10,
  },
  showlegend: false,
  title: '',
  width: 400,
  xaxis: {
    range: [30],
  },
  yaxis: {
    range: [30],
  },
};

export const defaultPlotlyChartProps = {
  config: {} as Partial<Plotly.Config>,
  data: [] as Array<Partial<IPlotlyData>>,
  layout: {} as Partial<Plotly.Layout>,
  onClickCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
  onHoverCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
  onSelectedCallback: (event: plotly.PlotSelectionEvent) => {
    return;
  },
  onUnHoverCallback: (event: plotly.PlotMouseEvent) => {
    return;
  },
};

export type PlotlyChartProps = {
  data: Array<Partial<IPlotlyData>>;
} & Partial<typeof defaultPlotlyChartProps>;

/**
 * React wrapper for a Plotly Chart.
 *
 * @description
 * Based upon: https://github.com/davidctj/react-plotlyjs-ts
 *
 * @export
 * @extends {React.Component<IPlotlyChartProps, any>}
 */
export class PlotlyChartClass extends React.Component<PlotlyChartProps, any> {
  public plotlyCanvas: plotly.PlotlyHTMLElement | null = null;
  protected canvasRef: HTMLDivElement | null = null;

  public attachListeners() {
    this.plotlyCanvas!.on('plotly_click', this.onClick);
    this.plotlyCanvas!.on('plotly_selected', this.onSelect);

    this.plotlyCanvas!.on('plotly_hover', this.onHover);
    this.plotlyCanvas!.on('plotly_unhover', this.onUnHover);

    window.removeEventListener('resize', this.resize);
    window.addEventListener('resize', this.resize);
  }

  public resize = () => {
    plotly.Plots.resize(this.plotlyCanvas!);
  };

  public draw = async () => {
    const { data, layout, config } = this.props;
    if (this.plotlyCanvas) {
      // plotly.react will not destroy the old plot: https://plot.ly/javascript/plotlyjs-function-reference/#plotlyreact
      this.plotlyCanvas = await plotly.react(
        this.plotlyCanvas,
        data,
        this.getMergedLayout(layout),
        this.getMergedConfig(config),
      );
    }
  };

  public componentDidUpdate(prevProps: PlotlyChartProps) {
    const { data, layout, config } = this.props;
    if (data !== prevProps.data || layout !== prevProps.layout || config !== prevProps.config) {
      this.draw();
    }
  }

  public async componentDidMount() {
    if (this.canvasRef && !this.plotlyCanvas) {
      const { data, layout, config } = this.props;
      this.plotlyCanvas = await plotly.react(
        this.canvasRef,
        data,
        this.getMergedLayout(layout),
        this.getMergedConfig(config),
      );
      this.attachListeners();
      this.draw();
    }
  }

  public componentWillUnmount() {
    if (this.plotlyCanvas) {
      plotly.purge(this.plotlyCanvas);
      this.plotlyCanvas = null;
    }
    window.removeEventListener('resize', this.resize);
  }

  public render() {
    return (
      <div className={'plotly-chart'} ref={node => (this.canvasRef = node ? node : null)} style={{ marginBottom: 5 }} />
    );
  }

  protected getMergedConfig = (config: Partial<Plotly.Config> = {}) => {
    const foo: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, defaultPlotlyConfig));
    const bar: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, config));
    return Object.assign({}, foo.mergeDeep(bar).toJS());
  };

  protected getMergedLayout = (layout: Partial<Plotly.Layout> = {}) => {
    const foo: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, defaultPlotlyLayout));
    const bar: Immutable.Map<any, any> = Immutable.fromJS(Object.assign({}, layout));
    return Object.assign({}, foo.mergeDeep(bar).toJS());
  };

  protected onClick = (event: plotly.PlotMouseEvent) => {
    const { onClickCallback } = this.props;
    if (onClickCallback) {
      onClickCallback(event);
    }
  };

  protected onHover = (event: plotly.PlotMouseEvent) => {
    const { onHoverCallback } = this.props;
    if (onHoverCallback) {
      onHoverCallback(event);
    }
  };

  protected onSelect = (event: plotly.PlotSelectionEvent) => {
    const { onSelectedCallback } = this.props;
    if (onSelectedCallback) {
      onSelectedCallback(event);
    }
  };

  protected onUnHover = (event: plotly.PlotMouseEvent) => {
    const { onUnHoverCallback } = this.props;
    if (onUnHoverCallback) {
      onUnHoverCallback(event);
    }
  };
}

const PlotlyChart = withDefaultProps(defaultPlotlyChartProps, PlotlyChartClass);

export default PlotlyChart;
export { PlotlyChart };
