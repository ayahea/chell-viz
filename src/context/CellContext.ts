import { CELL_TYPE } from 'chell';
import * as React from 'react';

export const initialCellContext = {
  addCells: (cells: CELL_TYPE[]) => {
    return;
  },
  currentCells: [] as CELL_TYPE[],
  removeAllCells: (cells: CELL_TYPE[]) => {
    return;
  },
  removeCells: (cells: CELL_TYPE[]) => {
    return;
  },
};

export const CellContext = React.createContext({
  ...initialCellContext,
});
