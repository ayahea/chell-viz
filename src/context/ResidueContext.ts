import { RESIDUE_TYPE } from 'chell';
import * as React from 'react';

export interface IResidueSelection {
  [key: string]: RESIDUE_TYPE[];
}

export interface IResidueContext {
  addCandidateResidue: (residue: RESIDUE_TYPE) => void;
  addLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
  candidateResidue: RESIDUE_TYPE | 'none';
  lockedResiduePairs: IResidueSelection;
  removeAllLockedResiduePairs: () => void;
  removeCandidateResidue: () => void;
  removeLockedResiduePair: (residues: RESIDUE_TYPE[]) => void;
}

export const initialResidueContext = {
  candidateResidue: 'none' as RESIDUE_TYPE | 'none',
  lockedResiduePairs: {} as IResidueSelection,
} as IResidueContext;

export const ResidueContext = React.createContext({
  ...initialResidueContext,
});
