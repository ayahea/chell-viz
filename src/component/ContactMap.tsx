import * as React from 'react';

import { IContactMapData, ICouplingScore, RESIDUE_TYPE } from '../../types/chell';
import ResidueContext, { IResidueSelection } from '../context/ResidueContext';
import PlotlyChart, { defaultConfig, defaultLayout, generatePointCloudData } from '../helper/PlotlyHelper';
import { withDefaultProps } from '../helper/ReactHelper';
import ChellSlider from './ChellSlider';

export type CONTACT_MAP_CB_RESULT_TYPE = ICouplingScore;
export type ContactMapCallback = (coupling: CONTACT_MAP_CB_RESULT_TYPE) => void;

const defaultProps = {
  contactColor: '#009999',
  couplingColor: '#000000',
  data: {
    contactMonomer: [],
    couplingScore: [],
    distanceMapMonomer: [],
  } as IContactMapData,
  enableSliders: false,
  height: 400,
  highlightColor: '#0000ff',
  onClick: undefined as ContactMapCallback | undefined,
  onMouseEnter: undefined as ContactMapCallback | undefined,
  padding: 0,
  selectedData: undefined as number | undefined,
  width: 400,
};

const initialState = {
  contactPoints: new Float32Array(0),
  couplingPoints: new Float32Array(0),
  highlightedPoints: new Float32Array(0),
  nodeSize: 4,
  probabilityFilter: 0.99,
};

type Props = {} & typeof defaultProps;
type State = Readonly<typeof initialState>;

export class ContactMapComponent extends React.Component<Props, State> {
  public readonly state: State = initialState;

  constructor(props: Props) {
    super(props);
  }

  public componentDidMount() {
    const { data } = this.props;
    if (data) {
      this.setupData(data);
    }
  }

  public componentDidUpdate(prevProps: Props, prevState: State) {
    const { data } = this.props;
    const isFreshDataView = data !== prevProps.data || this.state.probabilityFilter !== prevState.probabilityFilter;
    if (isFreshDataView) {
      this.setupData(data);
    }
  }

  public render() {
    const { contactColor, couplingColor, height, highlightColor, padding, width } = this.props;
    const { contactPoints, couplingPoints } = this.state;

    return (
      <ResidueContext.Consumer>
        {({
          addLockedResiduePair,
          addHoveredResidues,
          candidateResidues,
          hoveredResidues,
          lockedResiduePairs,
          removeLockedResiduePair,
        }) => (
          <div id="ContactMapComponent" style={{ padding }}>
            <PlotlyChart
              config={{
                ...defaultConfig,
              }}
              data={[
                generatePointCloudData(contactPoints, contactColor, this.state.nodeSize),
                generatePointCloudData(couplingPoints, couplingColor, this.state.nodeSize),
                generatePointCloudData(
                  this.getHighlightedResidues(lockedResiduePairs),
                  highlightColor,
                  this.state.nodeSize,
                ),
              ]}
              layout={{
                ...defaultLayout,
                height,
                width,
                xaxis: {
                  ...defaultLayout.xaxis,
                  gridcolor: '#ff0000',
                  gridwidth: this.state.nodeSize,
                  showticklabels: false,
                  tickvals: [...candidateResidues, ...hoveredResidues],
                },
                yaxis: {
                  ...defaultLayout.yaxis,
                  gridcolor: '#ff0000',
                  gridwidth: this.state.nodeSize,
                  showticklabels: false,
                  tickvals: [...candidateResidues, ...hoveredResidues],
                },
              }}
              onHoverCallback={this.onMouseEnter(addHoveredResidues)}
              onClickCallback={this.onMouseClick(addLockedResiduePair)}
              onSelectedCallback={this.onMouseSelect()}
            />
            {this.props.enableSliders && this.renderSliders()}
          </div>
        )}
      </ResidueContext.Consumer>
    );
  }

  protected renderSliders() {
    const { width } = this.props;
    const sliderStyle = { width };
    return (
      <div>
        <ChellSlider
          max={100}
          min={0}
          label={'Probability'}
          defaultValue={99}
          onChange={this.onProbabilityChange()}
          style={sliderStyle}
        />
        <ChellSlider
          max={5}
          min={1}
          label={'Node Size'}
          defaultValue={this.state.nodeSize}
          onChange={this.onNodeSizeChange()}
          style={sliderStyle}
        />
      </div>
    );
  }

  protected setupData(data: IContactMapData) {
    const blackDots = new Array<ICouplingScore>();
    data.couplingScore.filter(coupling => coupling.probability >= this.state.probabilityFilter).forEach(coupling => {
      blackDots.push(coupling);
      blackDots.push({
        ...coupling,
        i: coupling.j,
        // tslint:disable-next-line:object-literal-sort-keys
        A_i: coupling.A_j,
        j: coupling.i,
        A_j: coupling.A_i,
      });
    });

    const contactPoints = new Float32Array(data.contactMonomer.length * 2);
    data.contactMonomer.forEach((contact, index) => {
      contactPoints[index * 2] = contact.i;
      contactPoints[index * 2 + 1] = contact.j;
    });

    const couplingPoints = new Float32Array(blackDots.length * 2);
    blackDots.forEach((coupling, index) => {
      couplingPoints[index * 2] = coupling.i;
      couplingPoints[index * 2 + 1] = coupling.j;
    });

    this.setState({
      contactPoints,
      couplingPoints,
    });
  }

  protected getHighlightedResidues(pairs: IResidueSelection): Float32Array {
    const pairKeys = Object.keys(pairs);
    const highlightedPoints: number[] = [];
    for (const key of pairKeys) {
      for (const residue of pairs[key]) {
        highlightedPoints.push(residue);
      }
    }
    return new Float32Array([...highlightedPoints, ...highlightedPoints.slice().reverse()]);
  }

  protected onClick = () => (coupling: ICouplingScore) => {
    if (this.props.onClick) {
      this.props.onClick(coupling);
    }
  };

  protected onProbabilityChange = () => (value: number) => {
    this.setState({
      probabilityFilter: value / 100,
    });
  };

  protected onNodeSizeChange = () => (value: number) => {
    this.setState({
      nodeSize: value,
    });
  };

  protected onMouseEnter = (cb: (residue: RESIDUE_TYPE[]) => void) => (e: Plotly.PlotMouseEvent) => {
    const { points } = e;
    cb([points[0].x, points[0].y]);
  };

  protected onMouseClick = (cb: (residues: RESIDUE_TYPE[]) => void) => (e: Plotly.PlotMouseEvent) => {
    const { points } = e;
    cb([points[0].x, points[0].y]);
  };

  protected onMouseSelect = () => (e: Plotly.PlotSelectionEvent) => {
    console.log(`onMouseSelect: ${e}`);
  };
}

const ContactMap = withDefaultProps(defaultProps, ContactMapComponent);

export default ContactMap;
export { ContactMap };
