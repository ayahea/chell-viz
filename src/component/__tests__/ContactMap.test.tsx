import { CommonWrapper, mount, ReactWrapper, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';
import * as Renderer from 'react-test-renderer';

import { initialResidueContext, IResidueContext } from '../../context/ResidueContext';
import { ICouplingScore } from '../../data/chell-data';
import { PlotlyChartClass } from '../chart/PlotlyChart';
import ContactMap, { ContactMapClass, ContactMapProps } from '../ContactMap';

// https://medium.com/@ryandrewjohnson/unit-testing-components-using-reacts-new-context-api-4a5219f4b3fe
// Provides a dummy context for unit testing purposes.
const getComponentWithContext = (context: IResidueContext = { ...initialResidueContext }) => {
  jest.doMock('../../context/ChellContext', () => {
    return {
      ChellContext: {
        Consumer: (props: any) => props.children(context),
      },
    };
  });

  return require('../ContactMap');
};

/**
 * Helper function to create and wait for a ContactMap to be mounted.
 *
 * @param props Custom props to be passed to the chart.
 * @returns A wrapper for the ContactMap that has been mounted.
 */
const getMountedContactMap = async (props?: Partial<ContactMapProps>) => {
  const Component = getComponentWithContext();
  const wrapper = mount(<Component.ContactMapClass {...props} />);
  await wrapper.mount();
  return wrapper;
};

/**
 * Helper function to create and wait for a shallow ContactMap.
 *
 * @param props Custom props to be passed to the chart.
 * @returns A wrapper for the ContactMap that has been shallowly created.
 */
const getShallowContactMap = (props?: Partial<ContactMapProps>) => {
  const Component = getComponentWithContext();
  return shallow(<Component.ContactMapClass {...props} />);
};

/**
 * Helper function to dispatch an event through plotly.
 *
 * @param wrapper The PlotlyChart.
 * @param event The name of the event to dispatch.
 */
const dispatchPlotlyEvent = (wrapper: ReactWrapper, eventName: string) => {
  const plotlyWrapper = wrapper.find('PlotlyChartClass') as CommonWrapper;
  (plotlyWrapper.instance() as PlotlyChartClass).plotlyCanvas!.dispatchEvent(new Event(eventName));
};

describe('ContactMap', () => {
  beforeEach(() => {
    jest.resetModuleRegistry();
  });

  const emptyData = {
    couplingScores: [],
  };

  const generateCouplingScore = (
    i: number,
    j: number,
    dist: number,
    extra?: Partial<ICouplingScore>,
  ): ICouplingScore => ({
    dist,
    i,
    j,
    ...extra,
  });

  // Translated from example1/coupling_scores.csv
  const sampleCorrectPredictedContacts = [generateCouplingScore(56, 50, 2.4)];
  const sampleIncorrectPredictedContacts = [generateCouplingScore(42, 50, 20.4)];
  const sampleOutOfLinearDistContacts = [
    generateCouplingScore(45, 46, 1.3),
    generateCouplingScore(44, 45, 1.3),
    generateCouplingScore(56, 57, 1.3),
  ];
  const sampleObservedContacts = [...sampleCorrectPredictedContacts, generateCouplingScore(41, 52, 1.3)];

  const uniqueScores = new Set(
    Array.from([
      ...sampleCorrectPredictedContacts,
      ...sampleIncorrectPredictedContacts,
      ...sampleObservedContacts,
      ...sampleOutOfLinearDistContacts,
    ]),
  );

  const sampleData = {
    couplingScores: Array.from(uniqueScores),
  };

  describe('Snapshots', () => {
    test('Should match existing snapshot when given no data.', () => {
      expect(toJson(shallow(<ContactMap />))).toMatchSnapshot();
    });

    test('Should match existing snapshot when given empty data.', () => {
      expect(toJson(shallow(<ContactMap data={emptyData} />))).toMatchSnapshot();
    });

    test('Should match existing snapshot when given sample data and sliders are _not_ enabled.', () => {
      const component = Renderer.create(<ContactMap data={sampleData} enableSliders={false} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should match existing snapshot when given sample data and sliders are enabled.', () => {
      const component = Renderer.create(<ContactMap data={sampleData} enableSliders={true} />);
      expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should match snapshot when locked residues are added.', async () => {
      const wrapper = await getMountedContactMap({ data: sampleData });
      const expectedSelectedPoints = {
        '37,46': [37, 46],
        '8': [8],
      };
      wrapper.setProps({
        lockedResiduePairs: expectedSelectedPoints,
      });
      await wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  test('Should invoke callback to add locked residues when a click event is fired.', async () => {
    const onClickSpy = jest.fn();
    const wrapper = await getMountedContactMap({ addLockedResiduePair: onClickSpy, data: sampleData });
    dispatchPlotlyEvent(wrapper, 'plotly_click');

    expect(onClickSpy).toHaveBeenCalledTimes(1);
  });

  test('Should invoke callback to add hovered residues when a click event is fired.', async () => {
    const onHoverSpy = jest.fn();
    const wrapper = await getMountedContactMap({ addHoveredResidues: onHoverSpy, data: sampleData });
    dispatchPlotlyEvent(wrapper, 'plotly_hover');

    expect(onHoverSpy).toHaveBeenCalledTimes(1);
  });

  test('Should clear state if new data is given', async () => {
    const onClearResidueSpy = jest.fn();
    const wrapper = await getMountedContactMap({ clearAllResidues: onClearResidueSpy, data: sampleData });
    await wrapper.update();
    wrapper.setProps({
      data: emptyData,
    });
    await wrapper.update();
    expect(onClearResidueSpy).toHaveBeenCalledTimes(1);
  });

  describe('Configuration', () => {
    test('Should update the state when the configuration accordion is clicked', () => {
      const wrapper = getShallowContactMap({ data: sampleData, enableSliders: true });
      const instance = wrapper.instance() as ContactMapClass;
      const initialState = instance.state.showConfiguration;
      wrapper
        .find('.contact-map-configuration-toggle')
        .at(0)
        .simulate('click');
      expect(instance.state.showConfiguration).toBe(!initialState);
    });

    test('Should update node size when appropriate slider is updated.', () => {
      const wrapper = getShallowContactMap({ data: sampleData, enableSliders: true });
      const instance = wrapper.instance() as ContactMapClass;
      expect(instance.state.nodeSize).toBe(3);
      const expectedSize = 5;
      wrapper
        .find('.node-size-slider')
        .at(0)
        .simulate('change', expectedSize);
      expect(instance.state.nodeSize).toBe(expectedSize);
    });

    test('Should update number of predicted contacts to show when appropriate slider is updated.', () => {
      const wrapper = getShallowContactMap({ data: sampleData, enableSliders: true });
      const instance = wrapper.instance() as ContactMapClass;
      const expectedCount = 50;
      expect(instance.state.numPredictionsToShow).not.toBe(expectedCount);
      wrapper
        .find('.predicted-contact-slider')
        .at(0)
        .simulate('change', expectedCount);
      expect(instance.state.numPredictionsToShow).toBe(expectedCount);
    });

    test('Should update linear distance filter when appropriate slider is updated.', () => {
      const wrapper = getShallowContactMap({ data: sampleData, enableSliders: true });
      const instance = wrapper.instance() as ContactMapClass;
      const expected = 10;
      expect(instance.state.linearDistFilter).not.toBe(expected);
      wrapper
        .find('.linear-dist-filter')
        .at(0)
        .simulate('change', expected);
      expect(instance.state.linearDistFilter).toBe(expected);
    });

    test('Should update # of predicted contacts to show when appropriate slider is updated.', () => {
      const wrapper = getShallowContactMap({ data: sampleData, enableSliders: true });
      const instance = wrapper.instance() as ContactMapClass;
      const expected = 20;
      expect(instance.state.numPredictionsToShow).not.toBe(expected);
      wrapper
        .find('.predicted-contact-slider')
        .at(0)
        .simulate('change', expected);
      expect(instance.state.numPredictionsToShow).toBe(expected);
    });
  });
});
