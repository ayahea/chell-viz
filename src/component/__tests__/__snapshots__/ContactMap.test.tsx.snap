// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMap Snapshots Should match existing snapshot when given empty data. 1`] = `<[object Object] />`;

exports[`ContactMap Snapshots Should match existing snapshot when given no data. 1`] = `<[object Object] />`;

exports[`ContactMap Snapshots Should match existing snapshot when given sample data and sliders are _not_ enabled. 1`] = `
<div
  id="ContactMapComponent"
  style={
    Object {
      "padding": 0,
    }
  }
>
  <div
    className="plotly-chart"
    style={
      Object {
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  />
  <div />
</div>
`;

exports[`ContactMap Snapshots Should match existing snapshot when given sample data and sliders are enabled. 1`] = `
<div
  id="ContactMapComponent"
  style={
    Object {
      "padding": 0,
    }
  }
>
  <div
    className="plotly-chart"
    style={
      Object {
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  />
  <div />
  <div
    className="accordion ui fluid styled"
  >
    <div
      className="title contact-map-configuration-toggle"
      onClick={[Function]}
    >
      <i
        aria-hidden="true"
        className="dropdown icon"
      />
      Configuration
    </div>
    <div
      className="content"
    >
      <div
        className="node-size-slider"
        style={
          Object {
            "padding": 20,
          }
        }
      >
        <p>
          Node Size: 3
        </p>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "width": 360,
            }
          }
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "50%",
              }
            }
          />
          <div
            className="rc-slider-step"
          />
          <div
            aria-disabled={false}
            aria-valuemax={5}
            aria-valuemin={1}
            aria-valuenow={3}
            className="rc-slider-handle"
            role="slider"
            style={
              Object {
                "left": "50%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
      <div
        className="linear-dist-filter"
        style={
          Object {
            "padding": 20,
          }
        }
      >
        <p>
          Linear Distance Filter (|i - j|): 5
        </p>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "width": 360,
            }
          }
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "44.44444444444444%",
              }
            }
          />
          <div
            className="rc-slider-step"
          />
          <div
            aria-disabled={false}
            aria-valuemax={10}
            aria-valuemin={1}
            aria-valuenow={5}
            className="rc-slider-handle"
            role="slider"
            style={
              Object {
                "left": "44.44444444444444%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
      <div
        className="predicted-contact-slider"
        style={
          Object {
            "padding": 20,
          }
        }
      >
        <p>
          Top N Predictions to Show: 29
        </p>
        <div
          className="rc-slider"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "width": 360,
            }
          }
        >
          <div
            className="rc-slider-rail"
            style={Object {}}
          />
          <div
            className="rc-slider-track"
            style={
              Object {
                "left": "0%",
                "width": "48.275862068965516%",
              }
            }
          />
          <div
            className="rc-slider-step"
          />
          <div
            aria-disabled={false}
            aria-valuemax={59}
            aria-valuemin={1}
            aria-valuenow={29}
            className="rc-slider-handle"
            role="slider"
            style={
              Object {
                "left": "48.275862068965516%",
              }
            }
            tabIndex={0}
          />
          <div
            className="rc-slider-mark"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ContactMap Snapshots Should match snapshot when locked residues are added. 1`] = `
<ContactMapClass
  addCandidateResidues={[Function]}
  addHoveredResidues={[Function]}
  addLockedResiduePair={[Function]}
  candidateResidues={Array []}
  clearAllResidues={[Function]}
  correctColor="#ff0000"
  data={
    Object {
      "couplingScores": Array [
        Object {
          "dist": 2.4,
          "i": 56,
          "j": 50,
        },
        Object {
          "dist": 20.4,
          "i": 42,
          "j": 50,
        },
        Object {
          "dist": 1.3,
          "i": 41,
          "j": 52,
        },
        Object {
          "dist": 1.3,
          "i": 45,
          "j": 46,
        },
        Object {
          "dist": 1.3,
          "i": 44,
          "j": 45,
        },
        Object {
          "dist": 1.3,
          "i": 56,
          "j": 57,
        },
      ],
    }
  }
  enableSliders={false}
  height={400}
  highlightColor="#ff8800"
  hoveredResidues={Array []}
  incorrectColor="#000000"
  lockedResiduePairs={
    Object {
      "37,46": Array [
        37,
        46,
      ],
      "8": Array [
        8,
      ],
    }
  }
  observedColor="#0000ff"
  padding={0}
  removeAllLockedResiduePairs={[Function]}
  removeCandidateResidues={[Function]}
  removeHoveredResidues={[Function]}
  removeLockedResiduePair={[Function]}
  toggleLockedResiduePair={[Function]}
  width={400}
>
  <div
    id="ContactMapComponent"
    style={
      Object {
        "padding": 0,
      }
    }
  >
    <ContactMapChartClass
      candidateResidues={Array []}
      data={
        Array [
          Object {
            "hoverinfo": "x+y",
            "marker": Object {
              "colorscale": Array [
                Array [
                  0,
                  "#0000ff",
                ],
                Array [
                  1,
                  "rgb(100,177,200)",
                ],
              ],
            },
            "name": "Observed",
            "points": Array [
              Object {
                "dist": 2.4,
                "i": 56,
                "j": 50,
              },
              Object {
                "dist": 1.3,
                "i": 41,
                "j": 52,
              },
              Object {
                "dist": 1.3,
                "i": 45,
                "j": 46,
              },
              Object {
                "dist": 1.3,
                "i": 44,
                "j": 45,
              },
              Object {
                "dist": 1.3,
                "i": 56,
                "j": 57,
              },
            ],
          },
          Object {
            "hoverinfo": "x+y",
            "marker": Object {
              "color": "#000000",
            },
            "name": "Predicted Contact (57)",
            "points": Array [
              Object {
                "dist": 2.4,
                "i": 56,
                "j": 50,
              },
              Object {
                "dist": 20.4,
                "i": 42,
                "j": 50,
              },
              Object {
                "dist": 1.3,
                "i": 41,
                "j": 52,
              },
            ],
          },
          Object {
            "hoverinfo": "x+y",
            "marker": Object {
              "color": "#ff0000",
            },
            "name": "Correct Prediction",
            "points": Array [
              Object {
                "dist": 2.4,
                "i": 56,
                "j": 50,
              },
              Object {
                "dist": 1.3,
                "i": 41,
                "j": 52,
              },
            ],
          },
          Object {
            "marker": Object {
              "color": "#ff8800",
              "line": Object {
                "color": "#ff8800",
                "width": 3,
              },
              "symbol": "circle-open",
            },
            "name": "Selected Res. Pairs",
            "points": Array [
              Object {
                "i": 37,
                "j": 46,
              },
            ],
          },
        ]
      }
      dataTransformFn={[Function]}
      hoveredResidues={Array []}
      nodeSize={3}
      onClickCallback={[Function]}
      onHoverCallback={[Function]}
      onSelectedCallback={[Function]}
      onUnHoverCallback={[Function]}
      range={
        Array [
          0,
          62,
        ]
      }
    >
      <PlotlyChartClass
        config={Object {}}
        data={
          Array [
            Object {
              "hoverinfo": "none",
              "marker": Object {
                "color": Array [
                  2.4,
                  1.3,
                  1.3,
                  1.3,
                  1.3,
                  2.4,
                  1.3,
                  1.3,
                  1.3,
                  1.3,
                ],
                "colorscale": Array [
                  Array [
                    0,
                    "#0000ff",
                  ],
                  Array [
                    1,
                    "rgb(100,177,200)",
                  ],
                ],
                "size": 6,
              },
              "mode": "markers",
              "name": "Observed",
              "type": "scattergl",
              "x": Array [
                56,
                41,
                45,
                44,
                56,
                50,
                52,
                46,
                45,
                57,
              ],
              "y": Array [
                50,
                52,
                46,
                45,
                57,
                56,
                41,
                45,
                44,
                56,
              ],
              "z": Array [
                2.4,
                1.3,
                1.3,
                1.3,
                1.3,
                2.4,
                1.3,
                1.3,
                1.3,
                1.3,
              ],
            },
            Object {
              "hoverinfo": "none",
              "marker": Object {
                "color": "#000000",
                "size": 6,
              },
              "mode": "markers",
              "name": "Predicted Contact (57)",
              "type": "scattergl",
              "x": Array [
                56,
                42,
                41,
                50,
                50,
                52,
              ],
              "y": Array [
                50,
                50,
                52,
                56,
                42,
                41,
              ],
              "z": Array [
                2.4,
                20.4,
                1.3,
                2.4,
                20.4,
                1.3,
              ],
            },
            Object {
              "hoverinfo": "none",
              "marker": Object {
                "color": "#ff0000",
                "size": 6,
              },
              "mode": "markers",
              "name": "Correct Prediction",
              "type": "scattergl",
              "x": Array [
                56,
                41,
                50,
                52,
              ],
              "y": Array [
                50,
                52,
                56,
                41,
              ],
              "z": Array [
                2.4,
                1.3,
                2.4,
                1.3,
              ],
            },
            Object {
              "hoverinfo": "none",
              "marker": Object {
                "color": "#ff8800",
                "line": Object {
                  "color": "#ff8800",
                  "width": 3,
                },
                "size": 6,
                "symbol": "circle-open",
              },
              "mode": "markers",
              "name": "Selected Res. Pairs",
              "type": "scattergl",
              "x": Array [
                37,
                46,
              ],
              "y": Array [
                46,
                37,
              ],
              "z": Array [
                undefined,
                undefined,
              ],
            },
          ]
        }
        layout={
          Object {
            "legend": Object {
              "orientation": "h",
            },
            "showlegend": true,
            "xaxis": Object {
              "autorange": false,
              "range": Array [
                0,
                62,
              ],
            },
            "yaxis": Object {
              "autorange": false,
              "range": Array [
                62,
                0,
              ],
            },
          }
        }
        onClickCallback={[Function]}
        onHoverCallback={[Function]}
        onSelectedCallback={[Function]}
        onUnHoverCallback={[Function]}
      >
        <div
          className="plotly-chart"
          style={
            Object {
              "marginBottom": 10,
              "marginTop": 10,
            }
          }
        />
      </PlotlyChartClass>
    </ContactMapChartClass>
    <div />
  </div>
</ContactMapClass>
`;
